<launch>

  <arg name="GUI" default="true" />

  <include file="$(find mercury_panda)/launch/mask_rcnn_instance_segmentation.launch">
    <arg name="GUI" value="false" />
  </include>

  <group ns="camera">
    <node name="render_voxel_grids"
          pkg="morefusion_panda_ycb_video" type="render_voxel_grids.py"
          clear_params="true">
    </node>
    <node name="octomap_server"
          pkg="morefusion_panda_ycb_video" type="octomap_server"
          clear_params="true"
          output="screen">
      <remap from="~render" to="render_voxel_grids/render" />
      <remap from="~input/camera_info" to="color/camera_info" />
      <remap from="~input/depth" to="aligned_depth_to_color/image_raw" />
      <remap from="~input/points" to="depth_registered/points" />
      <remap from="~input/label_ins" to="mask_rcnn_instance_segmentation/output/label_ins" />
      <remap from="~input/class" to="mask_rcnn_instance_segmentation/output/class" />
      <rosparam subst_value="true">
        frame_id: panda_link0
        resolution: 0.01
        ground_as_noentry: false
      </rosparam>
    </node>
    <node name="grids_to_markers"
          pkg="morefusion_panda_ycb_video" type="voxel_grids_to_markers.py"
          clear_params="true">
      <remap from="~input" to="octomap_server/output/grids" />
      <rosparam subst_value="true">
        show_bbox: false
      </rosparam>
    </node>
    <group ns="octomap_server">
      <node name="draw_instance_segmentation"
            pkg="morefusion_panda_ycb_video" type="draw_instance_segmentation.py"
            clear_params="true">
        <remap from="~input/rgb" to="/camera/color/image_rect_color_passthrough/output" />
        <remap from="~input/label_ins" to="output/label_rendered" />
        <remap from="~input/class" to="output/class" />
        <rosparam subst_value="true">
          approximate_sync: true
        </rosparam>
      </node>
    </group>
  </group>

  <node name="singleview_3d_pose_estimation"
        pkg="morefusion_panda_ycb_video" type="singleview_3d_pose_estimation.py"
        clear_params="true"
        output="screen">
    <remap from="~input/camera_info" to="/camera/color/camera_info" />
    <remap from="~input/rgb" to="/camera/color/image_rect_color" />
    <remap from="~input/depth" to="/camera/aligned_depth_to_color/image_raw" />
    <remap from="~input/label_ins" to="/camera/octomap_server/output/label_rendered" />
    <remap from="~input/class" to="/camera/octomap_server/output/class" />
    <remap from="~input/grids_noentry" to="/camera/octomap_server/output/grids_noentry" />
    <rosparam subst_value="true">
      with_occupancy: true
    </rosparam>
  </node>
  <node name="object_poses_to_markers"
        pkg="morefusion_panda_ycb_video" type="object_poses_to_markers.py"
        clear_params="true">
    <remap from="~input" to="singleview_3d_pose_estimation/output" />
  </node>
  <node name="render_mesh_markers"
        pkg="morefusion_panda_ycb_video" type="render_mesh_markers.py"
        clear_params="true">
    <remap from="~input/camera_info" to="/camera/color/camera_info" />
    <remap from="~input/markers" to="object_poses_to_markers/output" />
  </node>
  <group ns="render_mesh_markers">
    <node name="tile_image"
          pkg="morefusion_panda_ycb_video" type="tile_image.py"
          clear_params="true">
      <rosparam subst_value="true">
        input_topics:
          - /camera/color/image_rect_color
          - output
        queue_size: 100
        approximate_sync: false
      </rosparam>
    </node>
  </group>

  <node name="object_mapping"
        pkg="morefusion_panda_ycb_video" type="object_mapping.py"
        clear_params="true">
    <remap from="~input/poses" to="singleview_3d_pose_estimation/output" />
    <remap from="~input/grids" to="/camera/octomap_server/output/grids" />
    <rosparam subst_value="true">
      frame_id: panda_link0
    </rosparam>
  </node>
  <group ns="object_mapping">
    <node name="object_poses_to_markers"
          pkg="morefusion_panda_ycb_video" type="object_poses_to_markers.py"
          clear_params="true">
      <remap from="~input" to="output/poses" />
    </node>
    <node name="render_mesh_markers"
          pkg="morefusion_panda_ycb_video" type="render_mesh_markers.py"
          clear_params="true">
      <remap from="~input/camera_info" to="/camera/color/camera_info" />
      <remap from="~input/markers" to="object_poses_to_markers/output" />
    </node>
    <group ns="render_mesh_markers">
      <node name="tile_image"
            pkg="morefusion_panda_ycb_video" type="tile_image.py"
            clear_params="true">
        <rosparam subst_value="true">
          input_topics:
            - /camera/color/image_rect_color
            - output
          queue_size: 100
          approximate_sync: false
        </rosparam>
      </node>
    </group>
  </group>

  <group if="$(arg GUI)">
    <node name="rviz"
          pkg="rviz" type="rviz"
          args="-d $(find mercury_panda)/rviz/morefusion.rviz"
          clear_params="true">
    </node>
  </group>

</launch>
